### PR TITLE
Features/refactoring

### PR DESCRIPTION
**Pull request by shawwn**
Howdy,

Would you like any of these changes?

- Rename `APPLY1` to `APPLY`
- Rename `NO`/`YES` to `FALSE?`/`TRUE?`
- Add tests for `NIL?`/`IS?`/`FALSE?`/`TRUE?`
- Replace each tab character with eight space characters
- Simplify `JOIN`
- Make `EXPAND-MACRO` local to `compiler.l`
- Modify `FOR` to evaluate the end condition once per iteration
- Modify `STEP` to support lists that are destructively modified during iteration

The changes to `FOR` and `STEP` allow iteration over e.g. lists that grow:

```
> (with l '(a b)
    (step x l
      (if (= x 'a) (add l 'c)
          (= x 'c) (add l 'd))))
("a" "b" "c" "d")
```

To verify that the `FOR`/`STEP` changes didn't introduce any slowdowns, I used `time` to measure how long it takes to run `make -B` three times for each lumen host (Node v6.1.0, Lua 5.2.4, and LuaJIT 2.1.0-beta2). Performance seems to be identical on all hosts.

Let me know which of the commits you dislike, and I'll submit another PR without those.

### CODE REVIEW

**Comment by sctb** [comment]
Let's drop the renaming of `yes` and `no`, and keep the others. 

It looks like renaming the `apply1` function to `apply` needs that `(do apply)` trick to get around non-terminating macroexpansion? It's curious that that works, though it would be better if it weren't needed. Couldn't we rewrite `apply1` to be variadic and drop the macro?

**Comment by shawwn** [comment]
> Let's drop the renaming of yes and no, and keep the others.

Ok, will do.

> It looks like renaming the apply1 function to apply needs that (do apply) trick to get around non-terminating macroexpansion?

Correct.

> Couldn't we rewrite `apply` to be variadic and drop the macro?

We could, but the `apply` function is a performance hotspot:

```
  obj/runtime.js
APPLY was called 1714 times
  obj/macros.js
APPLY was called 3714 times
  obj/main.js
APPLY was called 319 times
bin/lumen.js
bin/reader.js
APPLY was called 531 times
bin/compiler.js
APPLY was called 4027 times
bin/system.js
APPLY was called 179 times
  obj/runtime.lua
APPLY was called 1310 times
  obj/macros.lua
APPLY was called 3530 times
  obj/main.lua
APPLY was called 330 times
bin/lumen.lua
bin/reader.lua
APPLY was called 482 times
bin/compiler.lua
APPLY was called 3714 times
bin/system.lua
APPLY was called 152 times
  obj/test.js
APPLY was called 17222 times
  obj/test.lua
APPLY was called 17230 times
```

This would work, but it's terribly slow:

```
(define-global apply (f rest: args)
  (let args (stash (join (almost args) (last args)))
    (target js: ((get f 'apply) f args)
            lua: (f (values args)))))
```

Here's a solution with no performance overhead:

```
(define merge! (l)
  (join! l (drop l)))

(target js:
  (define-global apply (f)
    (let args (if (= arguments.length 2) arguments.1 (merge! (arguments% 1)))
      ((get f 'apply) f (stash args)))))

(target lua:
  (define-global apply (f |...|)
    (let args (if (= (select '# |...|) 1) (select 1 |...|) (merge! (list |...|)))
      (f (values (stash args))))))
```

It's verbose, but performance is identical to the existing codebase.

All my other attempts introduced significant slowdowns. (Total compile time ended up around 2% to 10% slower.)

So it seems like there are at least three options:

- Stick with the existing `apply` macro
- Use the above solution
- Drop support for variadic arguments by reverting to the old version of `apply`

What do you think?

For clarity, here's the diff of the proposed solution:

```diff
diff --git a/macros.l b/macros.l
index 7a7dec4..144d554 100644
--- a/macros.l
+++ b/macros.l
@@ -145,11 +145,6 @@
 (define-macro fn (args rest: body)
   `(%function ,@(bind* args body)))
 
-(define-macro apply (f rest: args)
-  (if (> (# args) 1)
-      `(apply1 ,f (join (list ,@(almost args)) ,(last args)))
-      `(apply1 ,f ,@args)))
-
 (define-macro guard (expr)
   (if (= target 'js)
       `((fn () (%try (list true ,expr))))
diff --git a/runtime.l b/runtime.l
index 4ef6e28..c0baf79 100644
--- a/runtime.l
+++ b/runtime.l
@@ -310,10 +310,18 @@
 (target lua:
   (define values (or unpack (get table 'unpack))))
 
-(define-global apply1 (f args)
-  (let args (stash args)
-    (target js: ((get f 'apply) f args)
-            lua: (f (values args)))))
+(define merge! (l)
+  (join! l (drop l)))
+
+(target js:
+  (define-global apply (f)
+    (let args (if (= arguments.length 2) arguments.1 (merge! (arguments% 1)))
+      ((get f 'apply) f (stash args)))))
+
+(target lua:
+  (define-global apply (f |...|)
+    (let args (if (= (select '# |...|) 1) (select 1 |...|) (merge! (list |...|)))
+      (f (values (stash args))))))
 
 (define-global call (f) (f))
```

**Comment by sctb** [comment]
Nice work! Let's keep your original patch for renaming `apply1` to `apply`. It's the smallest diff and keeps the runtime unchanged.

**Comment by shawwn** [comment]
Ok, ready to merge. Let me know if you think of anything else you'd like.

**Comment by shawwn** [comment]
Just to clarify, I pushed more commits to this PR yesterday. I think it contains all the changes you wanted, but if not, let me know.

**Comment by sctb** [comment]
Weird! I could've sworn I merged yesterday. Anyhow, it's in. Thank you!
